src/verifier/az_snp_vtpm/mod.rs

```diff
@@ -0,0 +1,143 @@
+// Copyright (c) Microsoft Corporation.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use super::{Attestation, TeeEvidenceParsedClaim, Verifier};
+use anyhow::{anyhow, Context, Result};
+use async_trait::async_trait;
+use az_snp_vtpm::certs::{get_chain_from_amd, get_vcek_from_amd};
+use az_snp_vtpm::hcl::{buf_to_hcl_data, HclReportWithRuntimeData};
+use az_snp_vtpm::report::Validateable;
+use az_snp_vtpm::vtpm::{Quote, VerifyVTpmQuote};
+use serde::{Deserialize, Serialize};
+use serde_json::json;
+use sha2::{Digest, Sha256, Sha384};
+use std::collections::BTreeMap;
+
+const HCL_VMPL_VALUE: u32 = 0;
+
+#[derive(Serialize, Deserialize)]
+struct Evidence {
+    quote: Quote,
+    report: Vec<u8>,
+}
+
+#[derive(Default)]
+pub struct AzSnpVtpm;
+
+#[async_trait]
+impl Verifier for AzSnpVtpm {
+    async fn evaluate(
+        &self,
+        nonce: String,
+        attestation: &Attestation,
+    ) -> Result<TeeEvidenceParsedClaim> {
+        let evidence = serde_json::from_str::<Evidence>(&attestation.tee_evidence)
+            .context("Failed to deserialize vTPM SEV-SNP evidence")?;
+
+        let hcl_report: HclReportWithRuntimeData = evidence.report[..].try_into()?;
+        let hashed_quote = hash_quote(attestation, &nonce);
+
+        verify_quote(&evidence.quote, &hcl_report, &hashed_quote)?;
+        verify_snp_report(&hcl_report)?;
+        verify_report_data(&hcl_report, &evidence.report)?;
+
+        let claim = parse_tee_evidence(&hcl_report);
+        Ok(claim)
+    }
+}
+
+fn verify_report_data(report: &HclReportWithRuntimeData, bytes: &[u8]) -> Result<()> {
+    let (_, var_data) = buf_to_hcl_data(bytes)?;
+
+    let mut hasher = Sha256::new();
+    hasher.update(var_data);
+    let var_data_hash = hasher.finalize().to_vec();
+
+    // Only the first 32 bytes of SNP report data are used for the sha256
+    let report_data = &report.snp_report().report_data[..32];
+
+    if var_data_hash != report_data {
+        return Err(anyhow!(
+            "SNP report data field is not matching runtime data"
+        ));
+    }
+
+    Ok(())
+}
+
+fn verify_quote(
```

Comment:
Would it make sense to make some of those functions (at least `verify_report_data`and `verify_snp_report` and probably `parse_tee_evidence`) `HclReportWithRuntimeData` methods in the upstream `az-snp-vtpm` crate?

Reply:
yes, indeed. the verification specifics should be pushed to the crate, if feasible. They're included here for illustrative purpose atm. regarding snp report verification: i'm hoping for a canonical implementation, in a virtee crate maybe.